data-mongodb/src/main/java/io/micronaut/data/mongodb/serde/DataDecoderContext.java

```diff
@@ -187,7 +188,7 @@ public <T> Deserializer<? extends T> findDeserializer(Argument<? extends T> type
         if (codec instanceof MappedCodec) {
             return ((MappedCodec<? extends T>) codec).deserializer;
         }
-        if (codec != null && !(codec instanceof IterableCodec)) {
+        if (codec != null && !(codec instanceof IterableCodec) && !(codec instanceof MapCodec)) {
```

Comment:
I have done a lot of debugging and this is the only fix I found. It might make sense, since there is also special case for `IterableCodec` as `MapCodec` deserialization deserializes object in the map as document and creates `BsonDocument` instead of actual class. This fix lets Micronaut Serde special (de)serializers do the work and fix the issue.